### PR TITLE
Feature/table controller

### PR DESCRIPTION
**Pull request by petropavel13**

### CODE REVIEW

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    // MARK: - Initialization
+
+    public init(style: UITableViewStyle, cellCreationType: CellCreationType = .OnTheFlight) {
+        self.cellCreationType = cellCreationType
+        super.init(style: style)
+    }
+
+    public required init?(coder aDecoder: NSCoder) {
+        if let cellCreationType = CellCreationType(rawValue: aDecoder.decodeIntegerForKey("CellCreationType")) {
```
Comment (scoreyou) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: `private static let creationTypeKey = "CellCreationType"`

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        fatalError("Your should implement cellIdentifierForIndexPath(_:)")
+    }
+
+    /**
+     method which configures cell before it can be used
+
+     - parameter cell:        UITableView or subclass cell
+     - parameter atIndexPath: index path of cell
+     */
+    public func configureCell(cell: UITableViewCell, atIndexPath: NSIndexPath) {
```
Comment (scoreyou) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: Is there ability to pass here casted to `CellType` instance?

Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: nope. you should use optional cast in subclasses like this:
```swift
switch cell {
case let labelCell as MyLabelCell:
  // do some thing with label cell
case let ticketCell as MyTicketCell:
  // do some things with ticketCell
default:
  // do nothing or call super.configureCell
}
```

File: LeadKit/LeadKit/Protocols/CellsControllerProtocol.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//  Copyright © 2016 Touch Instinct. All rights reserved.
+//
+
+import Foundation
+
+/**
+ *  protocol which ensures that specific type can configure cell and return cell attributes for specific index path
+ */
+public protocol CellsControllerProtocol {
+    associatedtype CellType
```
Comment (scoreyou) [inline LeadKit/LeadKit/Protocols/CellsControllerProtocol.swift]: Should not we use `typealias CellType = ...` in classes, which conforms this protocol?

Comment (petropavel13) [inline LeadKit/LeadKit/Protocols/CellsControllerProtocol.swift]: I don't get your question. See TableViewController example.

File: LeadKit/LeadKit/Classes/Cache/ViewsGenerator.swift [hunk 4]
```diff
@@ -19,9 +19,7 @@ public class ViewsGenerator<T where T: UIView>: ObjectsGenerator<T> {
      - returns: nothing
      */
     init(poolSize: UInt, nibName: String) {
-        super.init(poolSize: poolSize, objectsContructor: {() -> T in
-            T.loadFromNib(named: nibName)
-        })
+        super.init(poolSize: poolSize, objectsContructor: { T.loadFromNib(named: nibName) })
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Classes/Cache/ViewsGenerator.swift]: can you write `super(poolSize: poolSize, objectsContructor: { T.loadFromNib(named: nibName) })`?

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    private let cellCreationType: CellCreationType
+
+    private static let creationTypeKey = "CellCreationType"
+
+    // MARK: - Initialization
+
+    public init(style: UITableViewStyle, cellCreationType: CellCreationType = .OnTheFlight) {
+        self.cellCreationType = cellCreationType
+        super.init(style: style)
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: is it possible to write `super(style: style)`?

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    public required init?(coder aDecoder: NSCoder) {
+        if let cellCreationType = CellCreationType(rawValue: aDecoder.decodeIntegerForKey(TableViewController.creationTypeKey)) {
+            self.cellCreationType = cellCreationType
+            super.init(coder: aDecoder)
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: same?

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+//
+//  Created by Ivan Smolin on 30/05/16.
+//  Copyright © 2016 Touch Instinct. All rights reserved.
+//
+
+import UIKit
+
+public class TableViewController: UITableViewController, CellsControllerProtocol {
+
+    private var heightCache: [NSIndexPath: CGFloat] = [:]
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: Do you store heights of cells in ViewController? It's not good architecture, maybe it's better to store heights in cells view models?

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+            cell = tableView.dequeueReusableCellWithIdentifier(cellIdentifier) ?? cellsGenerator.get()
+        }
+
+        configureCell(cell, atIndexPath: indexPath)
+
+        return cell
+    }
+
+    // MARK: - Table view delegate
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: // MARK: - UITableViewDelegate

File: LeadKit/LeadKit/Enums/CellCreationType.swift [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+import Foundation
+
+/**
+ enum which describes cell creation behaviour of CellsControllerProtocol
+
+ - Preloaded:   cells is generated in advance
+ - OnTheFlight: cells is created on demand
+ */
+public enum CellCreationType: Int {
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Enums/CellCreationType.swift]: appple have example
``` obj-c
enum Planet: Int {
    case Mercury = 1, Venus, Earth, Mars, Jupiter, Saturn, Uranus, Neptune
}
```

File: LeadKit/LeadKit/Controllers/TableViewController.swift (hunk 5 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: nope

File: LeadKit/LeadKit/Controllers/TableViewController.swift (hunk 6 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: nope

File: LeadKit/LeadKit/Classes/Cache/ViewsGenerator.swift (hunk 4 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/Cache/ViewsGenerator.swift]: nope

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     method which adds cells generator for cells with specified reuse identifier
+
+     - parameter cellsGenerator: cells generator
+     - parameter cellIdentifier: cell reuse identifier
+     */
+    public func registerCellsGenerator(cellsGenerator: ViewsGenerator<UITableViewCell>,
+                                       forCellsWithIdentifier cellIdentifier: String) {
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: I think it's better to write this method in one line

Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: warning: Line Length Violation: Line should be 128 characters or less: currently 136 characters (line_length)

File: LeadKit/LeadKit/Extensions/UIView/UIView+DefaultReuseIdentifier.swift [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import UIKit
+
+extension UIView: ReuseIdentifierProtocol {
+
+    /**
+     default implementation of ReuseIdentifierProtocol
+
+     - returns: type name string
+     */
+    public class func reuseIdentifier() -> String {
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Extensions/UIView/UIView+DefaultReuseIdentifier.swift]: maybe it's better to use Read-Only Computed Property for this purpose

File: LeadKit/LeadKit/Classes/Cache/ViewsGenerator.swift (hunk 4 above)
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Classes/Cache/ViewsGenerator.swift]: oh, sorry

File: LeadKit/LeadKit/Enums/CellCreationType.swift (hunk 9 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Enums/CellCreationType.swift]: this style looks good for 8 values, and bad for only 2

Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Enums/CellCreationType.swift]:
``` swift
public enum CellCreationType: Int {

case Preloaded, OnTheFlight

}
```

Do you need the raw values?

File: LeadKit/LeadKit/Extensions/UIView/UIView+DefaultReuseIdentifier.swift (hunk 11 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Extensions/UIView/UIView+DefaultReuseIdentifier.swift]: nope. It will be impossible to override it in subclass

Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Extensions/UIView/UIView+DefaultReuseIdentifier.swift]: property is better than function, but it's not important, ok let's use method

File: LeadKit/LeadKit/Enums/CellCreationType.swift (hunk 9 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Enums/CellCreationType.swift]: I want to be sure that it will be the same always

Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Enums/CellCreationType.swift]: ok

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    // MARK: - Initialization
+
+    public init(style: UITableViewStyle, cellCreationType: CellCreationType = .OnTheFlight) {
+        self.cellCreationType = cellCreationType
+        super.init(style: style)
+    }
+
+    public required init?(coder aDecoder: NSCoder) {
+        if let creationType = CellCreationType(rawValue: aDecoder.decodeIntegerForKey(TableViewController.creationTypeKey)) {
+            self.cellCreationType = creationType
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: `cellCreationType = creationType`?

File: LeadKit/LeadKit/Controllers/TableViewController.swift [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+    // MARK: - Initialization
+
+    public init(style: UITableViewStyle, cellCreationType: CellCreationType = .OnTheFlight) {
+        self.cellCreationType = cellCreationType
+        super.init(style: style)
+    }
+
+    public required init?(coder aDecoder: NSCoder) {
+        if let creationType = CellCreationType(rawValue: aDecoder.decodeIntegerForKey(TableViewController.creationTypeKey)) {
```
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: guard in this case will looks better?

Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: I don't think so

File: LeadKit/LeadKit/Controllers/TableViewController.swift (hunk 12 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: what?

Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: `self` is not required in this case?

File: LeadKit/LeadKit/Controllers/TableViewController.swift (hunk 13 above)
Comment (fedorsemenchenko) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: one {block} instead of two)

Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: and we will have line length violation from linter. no, thanks :)

File: LeadKit/LeadKit/Controllers/TableViewController.swift (hunk 12 above)
Comment (petropavel13) [inline LeadKit/LeadKit/Controllers/TableViewController.swift]: now it's not required. self was used before variable rename

**Pull request merged**